mypyc/genops.py

```diff
@@ -2762,14 +2910,95 @@ def finalize_env_class(self) -> None:
 
     def instantiate_env_class(self) -> Value:
         """Assigns an environment class to a register named after the given function definition."""
-        self.fn_info.env_reg = self.add(Call(self.fn_info.env_class.ctor, [],
-                                             self.fn_info.fitem.line))
+        self.fn_info.callable_class.curr_env_reg = self.add(Call(self.fn_info.env_class.ctor, [],
+                                                                 self.fn_info.fitem.line))
 
         if self.fn_info.is_nested:
-            self.add(SetAttr(self.fn_info.env_reg, ENV_ATTR_NAME, self.fn_info.prev_env_reg,
+            self.add(SetAttr(self.fn_info.callable_class.curr_env_reg,
+                             ENV_ATTR_NAME,
+                             self.fn_info.callable_class.prev_env_reg,
                              self.fn_info.fitem.line))
 
-        return self.fn_info.env_reg
+        return self.fn_info.callable_class.curr_env_reg
+
+    def gen_generator_func(self) -> None:
+        self.setup_generator_class()
+        self.load_env_registers()
+        self.finalize_env_class()
+        self.add(Return(self.instantiate_generator_class()))
+
+    def setup_generator_class(self) -> ClassIR:
+        name = '{}_{}_gen'.format(self.fn_info.name, self.fn_info.ns)
+
+        generator_class = ClassIR(name, self.module_name)
+        generator_class.attributes[ENV_ATTR_NAME] = RInstance(self.fn_info.env_class)
+        generator_class.mro = [generator_class]
+
+        self.classes.append(generator_class)
+        self.fn_info.generator_class.ir = generator_class
+        return generator_class
+
+    def gen_iter_func(self, fn_info: FuncInfo) -> Tuple[List[BasicBlock], Environment, FuncInfo]:
+        self.enter(fn_info)
+        self_target = self.environment.add_local_reg(Var('self'),
+                                                     RInstance(fn_info.generator_class.ir),
+                                                     is_arg=True)
+        self.add(Return(self.read_from_target(self_target, fn_info.fitem.line)))
+        blocks, env, _, fn_info = self.leave()
+        return (blocks, env, fn_info)
+
+    def add_iter_to_generator_class(self,
+                                    blocks: List[BasicBlock],
+                                    env: Environment,
+                                    fn_info: FuncInfo) -> FuncIR:
+        sig = FuncSignature((RuntimeArg('self', object_rprimitive),), object_rprimitive)
+        iter_fn_decl = FuncDecl('__iter__', fn_info.generator_class.ir.name, self.module_name, sig)
+        iter_fn_ir = FuncIR(iter_fn_decl, blocks, env)
+        fn_info.generator_class.ir.methods['__iter__'] = iter_fn_ir
+        return iter_fn_ir
```

Comment:
I'm not sure if it is worth splitting these two functions? (At least, a bunch of other glue methods we generate we don't.)

Reply:
I separated it just so that it's easier to read from the `gen_func_item` function.